lib/plugins/amplitudePlugin.ts

```diff
@@ -0,0 +1,125 @@
+import { add, init, track } from '@amplitude/analytics-node'
+import type {
+  AmplitudeReturn,
+  BaseEvent,
+  NodeOptions,
+  Plugin,
+  Result,
+} from '@amplitude/analytics-types'
+import type {
+  FastifyInstance,
+  FastifyReply,
+  FastifyRequest,
+  HookHandlerDoneFunction,
+} from 'fastify'
+import fp from 'fastify-plugin'
+
+let pluginConfig: AmplitudeConfig | null = null
+
+async function plugin(
+  fastify: FastifyInstance,
+  config: AmplitudeConfig,
+  next: (err?: Error) => void,
+) {
+  pluginConfig = config
+
+  if (!config.isEnabled) {
+    return next()
+  }
+
+  if (!config.apiKey) {
+    return next(Error('Amplitude key not defined'))
+  }
+
+  await init(config.apiKey, config.options).promise
+
+  if (config.apiUsageTracking) {
+    enableApiUsageTracking(fastify, config.apiUsageTracking)
+  }
+  if (config.plugins) {
+    // @ts-expect-error
+    config.plugins.forEach((e) => add(e))
+  }
+
+  return next()
+}
+
+function enableApiUsageTracking(
+  fastify: FastifyInstance,
+  eventCreationFn: createApiTrackingEventFn,
+) {
+  fastify.addHook(
+    'onResponse',
+    (req: FastifyRequest, res: FastifyReply, done: HookHandlerDoneFunction) => {
+      const event = eventCreationFn(req, res)
+      if (event) {
+        amplitudeTrack(event)
+      }
+      return done()
+    },
+  )
+}
+
+/**
+ * Callback used to create the events that will be used to automatically
+ * track the API usage
+ */
+export type createApiTrackingEventFn = (req: FastifyRequest, res: FastifyReply) => BaseEvent | null
```

Comment:
🔶 By convention types should always start with an uppercase letter

Reply:
Nice catch! I will modify it right now, Thanks Rupert :D